File: 00165. Compare Version Numbers/solution.go
/**
 * Problem: 165. Compare Version Numbers
 *
 * Difficulty: Medium
 *
 * Language: Golang
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

func compareVersion(version1 string, version2 string) int {
	// Initialize pointers for version1 and version2
	i, j := 0, 0
	// Get lengths of version1 and version2
	n, m := len(version1), len(version2)

	// Loop until both pointers reach the end of their respective strings
	for i < n || j < m {
		// Initialize numbers for current revision of version1 and version2
		var revision1, revision2 int

		// Parse the current revision number from version1
		for i < n && version1[i] != '.' {
			revision1 = revision1*10 + int(version1[i]-'0')
			i++
		}

		// Skip the dot in version1 if present
		if i < n && version1[i] == '.' {
			i++
		}

		// Parse the current revision number from version2
		for j < m && version2[j] != '.' {
			revision2 = revision2*10 + int(version2[j]-'0')
			j++
		}

		// Skip the dot in version2 if present
		if j < m && version2[j] == '.' {
			j++
		}

		// Compare the current revision numbers
		if revision1 < revision2 {
			return -1
		}
		if revision1 > revision2 {
			return 1
		}
	}

	// All revisions are equal
	return 0
}

File: 00165. Compare Version Numbers/solution.js
/**
 * Problem: 165. Compare Version Numbers
 *
 * Difficulty: Medium
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Compares two version strings
 *
 * @param {string} version1 - First version
 * @param {string} version2 - Second version
 *
 * @returns {number} - 1 if version1 < version2, 1 if version1 > version2, 0 if equal
 */
const compareVersion = (version1, version2) => {
  // Split the version strings by '.' and convert each segment to a number
  const versionParts1 = version1.split('.').map(Number),
    versionParts2 = version2.split('.').map(Number)

  // Determine the maximum length between the two version arrays
  const maxLength = Math.max(versionParts1.length, versionParts2.length)

  // Iterate through each segment of the version numbers
  for (let i = 0; i < maxLength; i++) {
    // Get the current segment or default to 0 if undefined
    const segment1 = versionParts1[i] || 0,
      segment2 = versionParts2[i] || 0

    // Compare the current segments
    if (segment1 < segment2) return -1 // version1 is less than version2
    if (segment1 > segment2) return 1 // version1 is greater than version2
  }

  // All segments are equal
  return 0
}
